src/marin/post_training/rollout_worker.py

```diff
@@ -0,0 +1,332 @@
+        top_p: float | None = None,
+        top_k: int | None = None,
+    ) -> list[list[dict]]:
+        """Generate responses for a batch of prompts."""
+        host = self.inference_server.config.host
+        port = self.inference_server.config.port
+        url = f"http://{host}:{port}/v1/chat/completions"
+        responses = []
+        for prompt in prompts:
+            response = requests.post(
+                url,
+                headers={"Content-Type": "application/json"},
+                json={
+                    "model": getattr(self.inference_server.config, "model_name", "test-model"),
+                    "messages": [{"role": "user", "content": prompt}],
+                    "logprobs": True,
+                    "max_tokens": self.max_tokens,
+                    "temperature": temperature,
+                    "n": n_generations,
+                    "top_p": top_p,
+                    "top_k": top_k,
+                },
+            )
+            responses.append(response)
+
+        all_results = []
+        for response in responses:
+            if response.status_code != 200:
+                raise RuntimeError(f"Inference server error: {response.status_code} - {response.text}")
+
+            from openai.types.chat.chat_completion import ChatCompletion
+
+            chat_response = ChatCompletion.model_validate_json(response.text)
+            prompt_results = []
+            for choice in chat_response.choices:
+                content = choice.message.content
+                tokens = self._tokenizer.encode(content)
+                logprobs = [t.logprob for t in choice.logprobs.content]
+                prompt_results.append({"tokens": tokens, "logprobs": logprobs})
+            all_results.append(prompt_results)
+
+        return all_results
+
+    def compute_logprobs(
+        self,
+        input_tokens: np.ndarray,
+        input_attention_mask: np.ndarray,
+        target_tokens: np.ndarray,
+        target_attention_mask: np.ndarray,
+    ) -> np.ndarray:
+        """Compute log probabilities for given input/target pairs."""
+        import haliax as hax
+        from jax.nn import log_softmax
+
+        # Concatenate input and target tokens
+        full_tokens = jnp.concatenate([input_tokens, target_tokens], axis=1)
+        full_attention_mask = jnp.concatenate([input_attention_mask, target_attention_mask], axis=1)
+
+        # Convert to Haliax named arrays
+        # Get the axis names from the model config
+        Batch = hax.Axis("batch", full_tokens.shape[0])
+        SeqLen = hax.Axis("position", full_tokens.shape[1])
+
+        tokens_named = hax.named(full_tokens, (Batch, SeqLen))
+        attn_mask_named = hax.named(full_attention_mask.astype(jnp.bool_), (Batch, SeqLen))
+
+        # Compute logits using the model
+        # For causal language modeling, we need tokens[:-1] to predict tokens[1:]
+        input_tokens_named = tokens_named.slice(SeqLen, start=0, length=SeqLen.size - 1)
```

Comment:
if you want:

input_tokens_named = tokens_named[SeqLen, hax.ds(0, Seqlen.size - 1)]

Reply:
(ds is short for dynamic slice, which is shorthand in pallas too)